Fix QT4 build for mac

sethrj opened the pull request:
This PR:

- [x] adds support for the "latest" QT 4,
- [x] fixes an error in the `package` script that manifests on mac+clang builds,
- [x] fixes the build for High Sierra,
- [x] allows more components of QT to be disabled to improve chance of success
- [x] adds "+static" variant so it can be (in theory) built statically

Fixes #12139

scheibelp commented on var/spack/repos/builtin/packages/qt/package.py:
```diff
@@ -86,85 +100,53 @@ class Qt(Package):
-    # -webengine-webp            [system/qt] (Linux only)
-
+    else:
+        for _spec in ('%gcc','%intel'):
```
@sethrj there are a number of flake complaints, including no whitespace after the `,`

sethrj replied:
Fixed. Sorry!

sethrj commented:
Great! My bad regarding the flake issues (and multiple successive CI-triggering pushes on a PR). I'll get that done ASAP.

sethrj commented:
@michaelkuhn I had to make several more changes (the initial PR worked on the 10.13 SDK, then I added support for 10.14 SDK with clang-9, now it works with 10.14 SDK with clang-10).

I've also taken care of the flake8 warnings, sorry about that.

scheibelp commented on var/spack/repos/builtin/packages/glib/package.py:
```diff
@@ -53,6 +53,8 @@ class Glib(AutotoolsPackage):
     # around a legitimate usage.
     patch('no-Werror=format-security.patch')
+    # GCC8 on mac can preprocess obj-c headers but not build them
+    patch('configure-darwin.patch', when='platform=darwin')
```
Should this only apply for `gcc` or `gcc@8:`, based on the above comment?

Also to check: did you create this patch yourself? If not, it would be worth adding a link since that helps keep track of when it is no longer needed. (this applies to other patches added in the PR)

sethrj replied:
I created the patch myself. I believe it's more correct than what existed in the configure file -- the compiler *will* have to build the frameworks eventually, so checking to see that the carbon/cocoa headers exist may not be sufficient. I don't know if other GCC versions (or other non-apple compilers) can build on the mac without the patch, but I suspect not.

scheibelp replied:
>  the compiler will have to build the frameworks eventually, so checking to see that the carbon/cocoa headers exist may not be sufficient

True although I wonder if in general that kind of check would be hindered because the resources to truly build are not available at the time (so the cpp serves as a sanity check to resolve the includes).

Is this a similar issue to https://github.com/spack/spack/pull/10944#discussion_r270608085 where glib assumes the presence of some libraries based on include files but then cannot compile? If so, is there a reason you made the check more stringent here vs. the approach for python of adding the appropriate dependency?

sethrj replied:
No, the other case was a link failure. This failure is that the downstream code has valid syntax (preprocesses fine) but has *semantics* that are incompatible with GCC8:
```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/System/Library/Frameworks/Security.framework/Headers/Authorization.h:193:7: error: variably modified 'bytes' at file scope
  char bytes[kAuthorizationExternalFormLength];
```

scheibelp replied:
If I understand correctly, the affect is to disable these features (e.g. `glib_have_cocoa`) when building with gcc, so I would like to make sure: do you know if these are enabled when you build the package with apple's clang?

sethrj replied:
I just reconfigured with and without the patch with clang and gcc. Without the patch, clang successfully detected and built both cocoa and carbon. With the patch, clang only built carbon.
Without the patch, gcc configure says it can build carbon and cocoa, but the compiler errors when it includes the cocoa framework. With the patch, gcc configure says it can build neither carbon nor cocoa.

It sounds like I should change the patch command to
```py
    patch('configure-darwin.patch', when='%gcc platform=darwin')
```
to preserve the original behavior for the Apple clang compiler?

scheibelp replied:
> It sounds like I should change the patch command to...

@sethrj I agree. Based on your results, I'm guessing that `if ac_fn_c_try_compile` is too strict of a test for `cocoa` (whether clang or gcc is used) so you may want to relax that.

This brings up another point though: when it comes to reviewing Spack packages, I'd like to avoid reviewing novel patches for code when possible. Sometimes that's unavoidable but when that happens, if you could attach as much info as possible in the comment so other users of the package can track it that would be great:

* A link to any open issues regarding the failure (either in or outside of Spack)
* A description of the details of the system/compiler in used (e.g. including compiler version, even if the patch `when` clause is broader)
* The error message you get that this resolves

scheibelp commented on var/spack/repos/builtin/packages/python/package.py:
```diff
@@ -118,6 +118,7 @@ class Python(AutotoolsPackage):
     depends_on('tk', when='+tkinter')
     depends_on('tcl', when='+tkinter')
     depends_on('libuuid', when='+uuid')
+    depends_on('gettext')
```
Is this for providing gettext support in python? https://docs.python.org/2/library/gettext.html

sethrj replied:
When building `python@3.7.2%gcc@8.3.0`, the Python configure script found `libintl.h` (perhaps in the system paths somewhere?), but the corresponding `-lintl` inserted by configure failed because the responsible `gettext` installation wasn't in the library path. This was in the `_localemodule.c` file, but it looks like that code is indeed responsible for activating the `gettext` builtin module.

I'm actually unable to reproduce this failure now... so perhaps `gettext` should be an optional variant? Or should I just remove it for the time being?

scheibelp replied:
> I'm actually unable to reproduce this failure now... so perhaps gettext should be an optional variant? Or should I just remove it for the time being?

Based on the rest of your comment, `gettext` likely *should* be added as a dependency, since the python installation looks for it at some point (and uses a system version if it's available). The other option is to explicitly disable the python feature which requires `gettext` (it seems that most python installs have gotten by without it).

sethrj replied:
There's a part of the `configure` script with the comment "check if we need libintl for locale functions" and tries to link against a `textdomain` function; there's no option to disable it.

scheibelp replied:
OK, it sounds like `gettext` should in fact be added as a dependency then.

sethrj replied:
Sounds good. I also just verified that the header check for lintl is there in Python 2.7, so it seems like we shouldn't have to qualify the dependency with `when`.

adamjstewart replied:
Can you move this higher up under pkgconfig? This section is for optional dependencies. It's a shame there's no way to disable this dependency.

scheibelp commented on var/spack/repos/builtin/packages/qt/package.py:
```diff
@@ -86,84 +108,58 @@ class Qt(Package):
-    # -assimp            [system/qt/no]
-
-    # QtQml
-    depends_on("python", when='@5.7.0:', type='build')
```
This PR updates the `Python` package but at first glance appears to remove the python dependency. Is it pulled in transitively somewhere?

sethrj replied:
This was an unintentional deletion. Good catch!

sethrj commented:
@scheibelp After building some downstream code on my mac (which uses Fortran that links against wrapped QT code, ugh) I realized that there are more changes that need to be made to make the Mac QT4 install "ideal". Frankly, it would be ideal to "port" over [all the patches and configuration logic used by MacPorts](https://github.com/macports/macports-ports/tree/master/aqua/qt4-mac) -- what are the licensing implications of doing this?

sethrj commented:
@scheibelp Never mind about macports; I realized that all I needed to do to make the downstream happier was to install qt *not* as a framework. So I added the `+framework` variant: the default now is to install on all systems in `lib/` and `include/`, and to install in `Frameworks/` only when the variant is enabled. To restore the original behavior (install by default as a framework on mac) then we'll have to change the default from `False` to `sys.platform == 'darwin'`.

scheibelp commented on var/spack/repos/builtin/packages/qt/package.py:
```diff
@@ -69,15 +94,13 @@ class Qt(Package):
+    patch('qt4-pcre-include-conflict.patch', when='@4.8.6')
+    patch('qt4-tools.patch', when='@4+tools')
+    if not MAC_VERSION:
+        patch('qt4-gcc-and-webkit.patch', when='@4.8.6')
```
Could you also move the above comments to here vs. removing them?

scheibelp replied:
@sethrj do you think you would have a chance to do this in the next few days?

sethrj replied:
Yes, sorry I had a deadline and had to put the MR on hold. I'll get it done asap.

sethrj replied:
Yikes, when I said few days I truly didn't mean few months. I did restore the comments.

sethrj commented:
As per the contribution guidelines I've rebased the commits I can and squashed the more complex changes. @scheibelp Would you please take another look? There are no substantial changes since I addressed your last review feedback. Thanks!

adamjstewart commented on var/spack/repos/builtin/packages/qt/package.py:
```diff
@@ -91,84 +110,61 @@ class Qt(Package):
+        depends_on("xcb-util-wm")
+        depends_on("libxext", when='@3:4.99')
+        conflicts('+framework',
+                  msg="QT can only be build as a framework on macOS.")
```
typo: build -> built

sethrj replied:
Fixed.

sethrj commented:
Excellent! Yes, sorry it ended up being rather large -- it's basically a rewrite of the package definition for QT 4.

scheibelp commented:
Apologies but when I merged https://github.com/spack/spack/pull/12138 it generated a conflict here. Do you think you'll have an opportunity to sync it with develop in the near future?

sethrj commented:
@scheibelp No worries, rebased and fixed.

scheibelp commented:
I am not seeing the changes we discussed at https://github.com/spack/spack/pull/10944#discussion_r271775193 - did something change where you thought that was no longer necessary? To expand on https://github.com/spack/spack/pull/10944#discussion_r276360069, I think it would be more straightforward to explicitly disable `cocoa` by replacing

```
if ac_fn_c_try_compile "$LINENO"; then :
  glib_have_cocoa=yes
fi
```

with 

```
glib_have_cocoa=no
```

(since it appears the test fails even for compilers that support cocoa)

sethrj commented:
Sorry @scheibelp! I lost track of your previous comment. You're correct, even for mac clang it appears that it's still configuring with
```
checking for Mac OS X Cocoa support... no
```

Since that change to glib was because I was playing with building out a macOS GCC8 toolchain, and I no longer have that configuration set up, I'm just reverting the changes to glib. (Having these unrelated changes isn't exactly in line with the newish spack "one PR, one package" policy either.)

scheibelp commented on var/spack/repos/builtin/packages/qt/package.py:
```diff
@@ -70,15 +87,17 @@ class Qt(Package):
     # https://github.com/spack/spack/issues/1517
     patch('qt5-pcre.patch', when='@5:')
 
-    patch('qt4-corewlan-new-osx.patch', when='@4')
```
It doesn't appear that this patch is used anymore but the patch file itself was not removed.

Out of curiosity did the application of this patch cause the install to fail, or did you find it wasn't required (unlike `qt4-el-capitan.patch` it doesn't appear that this was combined into another patch file)?

sethrj replied:
Removing it now. It wasn't required (corewlan still appears to build), and the patch's changes would only affect Mac OS 10.8–13, and the current version is 10.15.

scheibelp commented on var/spack/repos/builtin/packages/qt/package.py:
```diff
@@ -70,15 +87,17 @@ class Qt(Package):
+        # recent versions of gcc.
+        # https://github.com/spack/spack/issues/9205
+        # https://github.com/spack/spack/issues/9209
+        patch('qt4-gcc-and-webkit.patch', when='@4.8.6')
```
Should this apply for `4.8.5` as well? You don't need to test if you don't know, but if you don't know then IMO it's worth applying this to both (since that's what was happening before).

sethrj replied:
I don't know, so I'll make that change.

scheibelp commented on var/spack/repos/builtin/packages/qt/package.py:
```diff
@@ -91,84 +110,55 @@ class Qt(Package):
+    if MACOS_VERSION is None:
+        depends_on("fontconfig")
+        depends_on("libx11")
+        depends_on("libxcb")
```
The original logic

```
depends_on("libxcb", when=sys.platform != 'darwin')
```

is a bit more idiomatic for Spack. This isn't particularly problematic though given that I assume people wouldn't be cross compiling this.

sethrj replied:
I thought for `depends_on`,  `when` had to be a spec rather than a bool? Also, how would
```python
    if MACOS_VERSION is None:
        depends_on("libxext", when='@3:4.99')
```
be represented?

scheibelp replied:
Good point, I was wrong about that. Also:

* I incorrectly stated in https://github.com/spack/spack/pull/12138#discussion_r308453777 that you cannot have `when=True` - you can.
* However, in your case this would not work because it is mixing spec evaluation with booleans which is not supported

So in short this is fine.

scheibelp commented on var/spack/repos/builtin/packages/qt/package.py:
```diff
@@ -44,12 +49,24 @@ class Qt(Package):
+    variant('ssl',    default=True,
+            description="Build with OpenSSL support.")
+    variant('freetype', default='system', description='Freetype2 support',
+            values=('system', 'qt', 'none'), multi=False)
```
Do you use `freetype=qt/none`, or have you been able to build with `freetype=system`? Ideally we could always force QT to use the Spack-built `freetype`.

I'll also note that while this translates directly to a QT configure option it is a confusing word choice for a Spack package: a "system" package in Spack often refers to a non-Spack package, where here it refers to a non-QT package that is built with Spack. The variant values `spack/qt/none` would be clearer IMO.

sethrj replied:
I always build with `freetype=none` to reduce the chances of another conflict between a contemporary library and the 15-year-old QT4 😭 

You're right the name choice is confusing outside the context of the QT configure options. I'll make the change.

scheibelp commented on var/spack/repos/builtin/packages/qt/package.py:
```diff
@@ -219,32 +209,89 @@ def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
+        }
+
+        def repl(match):
+            # Try to replace with our patch
```
IMO a better wording: "replace the original config variable value with the one chosen here if it is mentioned in 'patches'; otherwise return the original value.

sethrj replied:
Updated.

scheibelp commented:
Thanks!

sethrj commented:
Glad to! Thanks for the merge :)
